clients/cobol-lsp-vscode-extension/src/__tests__/services/copybook/download/CopybookDownloaderForE4E.spec.ts

```diff
@@ -0,0 +1,159 @@
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+  it("checks copybook downloaded into correct path", () => {
+    expect(
+      (CopybookDownloaderForE4E as any).getCopybookPath(
+        "Instance.Instance",
+        "pgm",
+        "/storagePath",
+        "Copy.cpy",
+      ),
+    ).toEqual(
+      path.join(
+        "/storagePath",
+        ".e4e",
+        ".copybooks",
+        "Instance.Instance",
+        "pgm",
+        "Copy",
+      ),
+    );
+    expect(fs.mkdirSync).toHaveBeenCalled();
+  });
+  it("checks not to try to download any if member or element not available in e4e", () => {
+    e4eDownloader.downloadCopybookE4E(
+      { name: "NoCopybook", dialect: "COBOL" },
+      "document-uri",
+      e4eResponseMock,
+    );
+
+    const spyDownloadDataset = jest.spyOn(e4eDownloader, "downloadDatasetE4E");
+    const spyDownloadElement = jest.spyOn(e4eDownloader, "downloadElementE4E");
+    expect(spyDownloadDataset).not.toHaveBeenCalled();
+    expect(spyDownloadElement).not.toHaveBeenCalled();
+  });
+  it("check download performed only for member when copyboook matches member and element", () => {
+    const spyDownloadDataset = jest.spyOn(e4eDownloader, "downloadDatasetE4E");
+    const spyDownloadElement = jest.spyOn(e4eDownloader, "downloadElementE4E");
+    e4eDownloader.downloadCopybookE4E(
+      { name: "copybook", dialect: "COBOL" },
+      "document-uri",
+      e4eResponseMock,
+    );
+    expect(spyDownloadDataset).toHaveBeenCalledWith(e4eResponseMock, {
+      dataset: "dataset",
+      member: "copybook",
+    });
+    expect(spyDownloadElement).not.toHaveBeenCalled();
+  });
+  it("check download performed only for element when no member matches", () => {
+    const spyDownloadDataset = jest.spyOn(e4eDownloader, "downloadDatasetE4E");
+    const spyDownloadElement = jest.spyOn(e4eDownloader, "downloadElementE4E");
+    const responseMock = Object.assign({}, e4eResponseMock);
+    responseMock.members = [];
+    e4eDownloader.downloadCopybookE4E(
+      { name: "copybook", dialect: "COBOL" },
+      "document-uri",
+      responseMock,
+    );
+    expect(spyDownloadElement).toHaveBeenCalledWith(
+      responseMock,
+      {
+        use_map: false,
+        environment: "environment",
+        stage: "stage",
+        system: "system",
+        subsystem: "subsystem",
+        type: "type",
+        element: "copybook",
+        fingerprint: "fingerprint",
+      },
+      "document-uri",
+    );
+    expect(spyDownloadDataset).not.toHaveBeenCalled();
+  });
+  it("check downloadDatasetE4E does not perform IO in case of Error", () => {
+    e4eDownloader.downloadCopybookE4E(
+      { name: "copybook", dialect: "COBOL" },
+      "document-uri",
+      e4eErrorResponseMock,
+    );
+    const spyGetMember = jest.spyOn(e4eErrorResponseMock.api, "getMember");
+    expect(spyGetMember).toHaveBeenCalled();
+    expect(fs.promises.writeFile).not.toHaveBeenCalled();
```

Comment:
This is again testing an implementation detail... There is so many other way to write a file, so checking this one particular way is meaningless.

Reply:
It is meaningful in this way ; If error occurs don't try to write to any file and don't perform any write operation.